_posts/2021-06-22-type-phantoms-and-safe-apis.md

```diff
@@ -0,0 +1,712 @@
+
+Usually, we'd also like meaningful error messages. In that case using `Result` with the `Validator`
+works well:
+
+```typescript
+type Validator<T> = (value: unknown) => Result<T, ErrorMessages<T>>
+```
+
+### Passing added context to handlers
+
+Often the server handler needs more data, not just the request parameters. If this data depends on
+the request, we need to initialize it before calling the handler. This may involve middleware to
+modify the request object. In that case, the server-side `Params` type would need changes. Another
+way is to pass a separate `context` object to the handlers:
+
+```typescript
+// lib/Server.ts
+
+type Context = { /* ... */ }
+
+declare function createContext(req: Request): Context
+
+export type Handler<EP extends Endpoint> =
+  (params: Params<EP>, context: Context) => Promise<EP['result']>
+
+const createHandler = <EP extends Endpoint>(handler: Handler<EP>): RouteHandler<EP> =>
+  async (req, res) => {
+    const context = createContext(req)
+
+    res.send(await handler({ path: req.params, body: req.body }, context))
+  }
+```
+
+If the *type* of `Context` also depends on the endpoint, the utilities would need changes as well.
+
+## Further improvements
+
+That's quite a few changes and tweaks already! Turns out it's not easy to write glue code and
+utilities that would be usable out of the box. Project requirements and techniques vary, and the
+glue code needs adjustments as well.
+
+We've not even touched on many common API features that may be needed:
+
+- How is authentication handled? What about authorization?
+- What about specific request and response headers?
+- How to make multiple separate APIs type safe?
+- Can we make route handling even more safe, for example with path uniqueness checks?
+
+Some of these are straightforward using techniques we already explored. That means even more changes
+to the common utilities though. Could we develop the core functionality further to make it more
+generic and support various kinds of API structures?
+
+### Adapting to different API structures
+
+Using custom adapters would help with some of these issues. The generic utilities are less useful
+then, but the request and response conversions can be customized. The client-side implementation
+could look like this:
+
+```typescript
+// lib/Client.ts
+
+type MakeRequest<EP extends Endpoint = Endpoint> =
+  (params: Omit<EP, 'result'>) => Promise<EP['result']>
+
+export function createHandlers<T extends Endpoints>(
+  makeRequest: MakeRequest,
+  routes: Routes<T>,
+): Handlers<T> {
+  /* ... */
+}
+```
+
+Now the common utility functions don't know the details of how to make requests, and those details
+can be tweaked freely. Some features still need type changes. For example, using the `Result` type
+means changing the `MakeRequest` type as well.
+
+The server implementation could use the same principles to extract library logic and enable
+customization.
+
+### Simplifying types
```

Comment:
If you want the shorten the post, I think that it could be possible to merge this part and the next one, in order to summarise some of the ways how the particular context of the app impact the approach. E.g. how the considerations might change from app A to app B.

I know that it's an "it depends" matter; but now it's a little ambiguous on _what it depends on_. (I think even a pointer or two would be helpful for painting a rough picture.)

Reply:
Hmm... I get the gist here, but so far haven't had any clear ideas on how to improve.

Reply:
No worries! I'm sure restructuring would be more effort than it's worth, and perhaps it wouldn't really shorten the post at all. I'm also having a really hard time trying to reword the point into something satisfying or useful; so I won't. 😄 

Essentially: the only real solution I've seen to the problem of _"it's context-dependent, but every context is different"_ is observing the potential invariants in the context based on past experiences. It's the primary reason why I've been **really** into patterns, specifically Alexandrian patterns ([like these](http://www.laputan.org/mud/mud.html)). I'm totally up for discussing this stuff some time, if you find it interesting.